### ISSUE TITLE
SodaCL column profiling

### ISSUE THREAD

**Issue by bastienboutonnet** [sodadata/soda-core#1256]
Generating and pushing profiling/EDA metrics to cloud will be done in soda-core via a profiling run.

The original proposition looks like this:

```yaml
profiling advanced:
  tables:
    - SODATEST_%
    - include SODATEST_%
    - exclude SODATEST_%
  columns:
    - TABLE%.COLNAME%
    - include TABLE%.COLNAME%
    - exclude TABLE%.COLNAME%
  max sample size: 100
  statistics: enabled  
```

We are however thinking about another top-level name along the lines of (based on a more explicit proposal from @janet-can:
```yaml
profile columns:
```

I assume the rest of the controls are going to stay the same meaning that users should configure column profiling via this canonical sodaCL spec:

```yaml
profile columns:
  tables: (required?)
    - prod_%
    - include prod_%
    - exclude dev_%
  columns: # not required
    - prod_%.some_column
    - prod_%.some_column_prefix%
    - include table.column
    - exclude table.column
  max sample size: 100 # not required and will default to a sensible number
  statistics: enabled
```

@tombaeyens can you confirm this makes sense language-wise? Also, can you explain the intention behind the `statistics` key? What does it control?

**Comment by tombaeyens** [comment]
Agreed.  Blessing given for changing the name to of the top level element to `profile column:`

**Comment by tombaeyens** [comment]
@bastienboutonnet It would be good to do a review on the proposal of the nested configuration parameters.  What are the options that all can be configured?  And how will they look like in SodaCL.  Can you do a proposal?

**Comment by bastienboutonnet** [comment]
Thanks for the questions @tombaeyens 

I can give further thoughts to what you ask. Before I do that I need some clarifications:
- what does "sample activation" mean?
- by "which metrics" I assume we may want to let users specifiy/override the metrics to be derived?
- "probably in groups or tshirt sizes or just on off switch" I don't really know what you mean. Can you rephrase a bit the way it fits in the sentence has me confused.

> should sample destination (Soda Cloud or S3) be configured under this column profiling section as well?

Am I correct in assuming you refer the sample data dataset tab? (screenshot below
<img width="1095" alt="Screenshot 2022-04-06 at 17 27 10" src="https://user-images.githubusercontent.com/4304794/162011172-85aafaad-cdf0-469e-8fb4-dd21a13f675a.png">
)

**Comment by tombaeyens** [comment]
> what does "sample activation" mean?
> Am I correct in assuming you refer the sample data dataset tab? (screenshot below

That is correct.  And sample activation just means turning on the sample creation in the SodaCL configuration.

> by "which metrics" I assume we may want to let users specifiy/override the metrics to be derived?

Yes.  The simplest approach would be just an on/off switch.  On would mean all the metrics needed to populate all the details for the column view in the UI.   We also could consider to add configuration to only calculate the basic metrics in an attempt to reduce the load on the DB while still having some metrics in the view.

I would suggest to only go for simple on/off switch first and only introduce more complex configs after we learn why customers would want more detailed configuration options.

**Comment by bastienboutonnet** [comment]
My thinking is based on the fact that we don't really know how users plan to want control over the calculated metrics that we don't offer the control for it here.

Based on the latest comment I propose the following SodaCL DSL

```yaml
profile columns:
  columns: 
    - prod_%.some_column
    - prod_%.some_column_prefix%
    - include table.column
    - exclude table.column
    - include table.%
    - include %.%
```

I'll open a new issue later where we can discuss how we give control over the metrics to be collected, and maybe involve some champions in that. What do you think @tombaeyens 

For now, this config will trigger all metrics that are part of profiling to be derived, on the selected columns.

**NOTE:**
- we have removed the concept of `tables:` since this DSL is about columns. However the use of wildcards (see above for some examples) allows users to include an entire table just as easily as they would have done so in the `tables:` section.

@tombaeyens I'd like your view on this:
I am thinking that the sample data capture, doesn't really fit into the `profile columns` concept but rather is better suited in the `table discovery` concept so I will amend the proposal made in #1257 to reflect that.

**Comment by bastienboutonnet** [comment]
One final question @tombaeyens. In soda-sql did we allow users to perform metric collection on a subset of the table (sample) or was it always done on the full table?

**Comment by tombaeyens** [comment]
> My thinking is based on the fact that we don't really know how users plan to want control over the calculated metrics that we don't offer the control for it here.

Correct: the main usage is to populate the EDA view.  Let's postpone all more fine grained control over the metrics for later.  No need to create an issue for that just yet.  I would suggest to let's the product team bring it up.  (as in: let's wait till customers ask for it)

> For now, this config will trigger all metrics that are part of profiling to be derived, on the selected columns.

Yes indeed.  That means all metrics needed to populate the columns EDA view.

> One final question @tombaeyens. In soda-sql did we allow users to perform metric collection on a subset of the table (sample) or was it always done on the full table?

I would be in favour of just having the feature on the full table first.  And then plan an extra iteration later to add the use of sampling with the SQL `TABLESAMPLE` thingy.

@mathissedestrooper is that ok for you?  We're proposing that the EDA columns view initially is done on the full table data.  This might be a problem for very large tables.  Those need some form of sampling.  What I propose is that users do not yet configure column profiling on those large tables till we add support for doing column profiling on samples of tables.   Is that ok for you?

**Comment by tombaeyens** [comment]
@bastienboutonnet @mathissedestrooper  Just think of something that I m not sure if we discussed already: invalid values for profiling columns.

Automatic type detection means that for text columns, we will match a sample against all the known valid format named regexes and see if a column matches the valid format regex.  

The reason that we did automatic type detection in Soda SQL on every scan was to get meaningful validity metrics in the column profiling (aka EDA) view.  Without this regex matching, all values would be valid.  Because there is no validity specified.

In Soda Core, I don't think we want to do this heavy matching each scan.  We only want to compute validity out of scan.  So as a user interaction.  But column profiling will not have the knowledge on user specified column types.  So that implies we won't have validity as part of the column profiling. (or in other words, it will always be 0% invalid)

Here we need to think what we want from a user experience. I can share more background.

**Comment by tombaeyens** [comment]
@bastienboutonnet There is the aspect of splitting the column profiling metrics from the metrics that are used for the checks.  Unlike Soda SQL, in Soda Core only column profiling metrics will populate the column view in the Soda Cloud UI in the columns view.  If there are metrics for checks should not interfere with the columns vizualization.

The JSON sent to the cloud should already reflect this.  Let me know if you encounter APIs / command JSONs that look different.  Also double check this behavior with Dirk on the backend.

**Comment by bastienboutonnet** [comment]
Ok so for this:
> Automatic type detection means that for text columns, we will match a sample against all the known valid format named regexes and see if a column matches the valid format regex.

I'm a bit confused with this. I don't see anywhere in the EDA views of cloud that we actually show valid/invalid values. So I'm not sure if I'm missing something.

And I'm not sure I follow this comment at all. Do you have more context?
> There is the aspect of splitting the column profiling metrics from the metrics that are used for the checks. Unlike Soda SQL, in Soda Core only column profiling metrics will populate the column view in the Soda Cloud UI in the columns view. If there are metrics for checks should not interfere with the columns vizualization.

The JSON sent to the cloud should already reflect this. Let me know if you encounter APIs / command JSONs that look different. Also double check this behavior with Dirk on the backend.

**Comment by tombaeyens** [comment]
> > Automatic type detection ...
> I'm a bit confused with this. 

I'll rephrase:

I think that going forward on Soda Core, column profiling *can not* produce invalid percentage or invalid count.  Simply because in the column profiling section there is no way to configure the expected validity rules for columns.  

If column profiling would need some form of invalid count or percentage number, then we need to define how that is going to be calculated and matched up with the other checks.

@mathissedestrooper does that work for you: that column details will not have an invalid count or percentage like in the screenshot below?

<img width="1056" alt="image" src="https://user-images.githubusercontent.com/944245/164507142-32ec44ba-84c2-4f9c-bf2c-0dcc242c78ac.png">

**Comment by bastienboutonnet** [comment]
Oh thanks @tombaeyens I had completely missed those fields! Yeah I would leave it out of this current version because it's not going to be easy to do that in soda-core I would say.

**Comment by mathissedestrooper** [comment]
Ok for me, just make sure the designs are reviewed by @soda-dimitar if we are removing fields

**Comment by bastienboutonnet** [comment]
@mathissedestrooper as far as I can see, if profiling doesn't push a metric it just doesn't show up. 

For example this dev table (https://dev.sodadata.io/datasets/d65820ba-8771-4d9d-a625-224f6be82dd5/columns) where I am currently testing profiling, just shows metrics that are provided to cloud but not the ones that are not.

That being said, we might want to revise the designs and perhaps the backend/frontend model (FYI @dirkgroenen) if we think we just won't ever provide that metric. But I don't think we need to do that now, because I'm pretty sure that if someone has a valid values check on a column then the metric of that specific check will actually be pushed to cloud and then it would show up so it would actually all be good. Not sure on that though, would have to be confirmed.

**Comment by bastienboutonnet** [comment]
Implemented
